--- a/dlls/winecoreaudio.drv/coreaudio.c
+++ b/dlls/winecoreaudio.drv/coreaudio.c
@@ -93,10 +93,9 @@
     BOOL playing, please_quit;
     REFERENCE_TIME period;
     UINT32 period_frames;
-    UINT32 bufsize_frames, resamp_bufsize_frames;
+    UINT32 bufsize_frames;
     UINT32 lcl_offs_frames, held_frames, wri_offs_frames, tmp_buffer_frames;
     UINT32 cap_bufsize_frames, cap_offs_frames, cap_held_frames;
-    UINT32 wrap_bufsize_frames;
     UINT64 written_frames;
     INT32 getbuf_last;
     WAVEFORMATEX *fmt;
@@ -432,21 +431,24 @@
 
     os_unfair_lock_lock(&stream->lock);
 
-    cap_wri_offs_frames = (stream->cap_offs_frames + stream->cap_held_frames) % stream->cap_bufsize_frames;
+    if(stream->cap_held_frames >= stream->cap_bufsize_frames - stream->cap_offs_frames)
+        cap_wri_offs_frames = stream->cap_held_frames -
+                              (stream->cap_bufsize_frames - stream->cap_offs_frames);
+    else
+        cap_wri_offs_frames = stream->cap_offs_frames + stream->cap_held_frames;
+
+    if(nframes > stream->cap_bufsize_frames){
+        os_unfair_lock_unlock(&stream->lock);
+        return kAudio_ParamError;
+    }
 
     list.mNumberBuffers = 1;
     list.mBuffers[0].mNumberChannels = stream->fmt->nChannels;
     list.mBuffers[0].mDataByteSize = nframes * stream->fmt->nBlockAlign;
 
-    if(!stream->playing || cap_wri_offs_frames + nframes > stream->cap_bufsize_frames){
-        if(stream->wrap_bufsize_frames < nframes){
-            free(stream->wrap_buffer);
-            stream->wrap_buffer = malloc(list.mBuffers[0].mDataByteSize);
-            stream->wrap_bufsize_frames = nframes;
-        }
-
+    if(!stream->playing || nframes > stream->cap_bufsize_frames - cap_wri_offs_frames)
         list.mBuffers[0].mData = stream->wrap_buffer;
-    }else
+    else
         list.mBuffers[0].mData = stream->cap_buffer + cap_wri_offs_frames * stream->fmt->nBlockAlign;
 
     sc = AudioUnitRender(stream->unit, flags, ts, bus, nframes, &list);
@@ -463,12 +465,18 @@
                     stream->wrap_buffer, list.mBuffers[0].mDataByteSize);
         }
 
-        stream->cap_held_frames += list.mBuffers[0].mDataByteSize / stream->fmt->nBlockAlign;
-        if(stream->cap_held_frames > stream->cap_bufsize_frames){
-            stream->cap_offs_frames += stream->cap_held_frames % stream->cap_bufsize_frames;
-            stream->cap_offs_frames %= stream->cap_bufsize_frames;
+        nframes = list.mBuffers[0].mDataByteSize / stream->fmt->nBlockAlign;
+        if(nframes > stream->cap_bufsize_frames - stream->cap_held_frames){
+            UINT32 overflow_frames = nframes - (stream->cap_bufsize_frames - stream->cap_held_frames);
+
+            if(overflow_frames >= stream->cap_bufsize_frames - stream->cap_offs_frames)
+                stream->cap_offs_frames = overflow_frames -
+                                          (stream->cap_bufsize_frames - stream->cap_offs_frames);
+            else
+                stream->cap_offs_frames += overflow_frames;
             stream->cap_held_frames = stream->cap_bufsize_frames;
-        }
+        }else
+            stream->cap_held_frames += nframes;
     }
 
     os_unfair_lock_unlock(&stream->lock);
@@ -865,6 +873,62 @@
         goto end;
     }
 
+    size = stream->bufsize_frames * stream->fmt->nBlockAlign;
+    if(NtAllocateVirtualMemory(GetCurrentProcess(), (void **)&stream->local_buffer, zero_bits,
+                               &size, MEM_COMMIT, PAGE_READWRITE)){
+        params->result = E_OUTOFMEMORY;
+        goto end;
+    }
+    silence_buffer(stream, stream->local_buffer, stream->bufsize_frames);
+
+    if(stream->flow == eCapture){
+        UInt32 max_frames, converter_input_bytes, converter_input_frames, property_size;
+
+        property_size = sizeof(max_frames);
+        sc = AudioUnitGetProperty(stream->unit, kAudioUnitProperty_MaximumFramesPerSlice,
+                                  kAudioUnitScope_Global, 0, &max_frames, &property_size);
+        if(sc != noErr){
+            WARN("Couldn't get maximum frames per slice: %x\n", (int)sc);
+            params->result = osstatus_to_hresult(sc);
+            goto end;
+        }
+
+        if(!stream->fmt->nBlockAlign || stream->period_frames > UINT_MAX / stream->fmt->nBlockAlign){
+            params->result = E_OUTOFMEMORY;
+            goto end;
+        }
+
+        converter_input_bytes = stream->period_frames * stream->fmt->nBlockAlign;
+        property_size = sizeof(converter_input_bytes);
+        sc = AudioConverterGetProperty(stream->converter, kAudioConverterPropertyCalculateInputBufferSize,
+                                       &property_size, &converter_input_bytes);
+        if(sc != noErr){
+            WARN("Couldn't get converter input buffer size: %x\n", (int)sc);
+            params->result = osstatus_to_hresult(sc);
+            goto end;
+        }
+        converter_input_frames = converter_input_bytes / stream->fmt->nBlockAlign +
+                                 !!(converter_input_bytes % stream->fmt->nBlockAlign);
+
+        stream->cap_bufsize_frames = muldiv(params->duration, stream->dev_desc.mSampleRate, 10000000);
+        stream->cap_bufsize_frames = max(stream->cap_bufsize_frames, max_frames);
+        stream->cap_bufsize_frames = max(stream->cap_bufsize_frames, converter_input_frames);
+
+        if(!stream->cap_bufsize_frames ||
+           stream->cap_bufsize_frames > UINT_MAX / stream->fmt->nBlockAlign){
+            params->result = E_OUTOFMEMORY;
+            goto end;
+        }
+
+        stream->cap_buffer = malloc(stream->cap_bufsize_frames * stream->fmt->nBlockAlign);
+        stream->wrap_buffer = malloc(stream->cap_bufsize_frames * stream->fmt->nBlockAlign);
+        stream->resamp_buffer = malloc(stream->period_frames * stream->fmt->nBlockAlign);
+        if(!stream->cap_buffer || !stream->wrap_buffer || !stream->resamp_buffer){
+            params->result = E_OUTOFMEMORY;
+            goto end;
+        }
+    }
+
     /* we play audio continuously because AudioOutputUnitStart sometimes takes
      * a while to return */
     sc = AudioOutputUnitStart(stream->unit);
@@ -874,24 +938,20 @@
         goto end;
     }
 
-    size = stream->bufsize_frames * stream->fmt->nBlockAlign;
-    if(NtAllocateVirtualMemory(GetCurrentProcess(), (void **)&stream->local_buffer, zero_bits,
-                               &size, MEM_COMMIT, PAGE_READWRITE)){
-        params->result = E_OUTOFMEMORY;
-        goto end;
-    }
-    silence_buffer(stream, stream->local_buffer, stream->bufsize_frames);
-
-    if(stream->flow == eCapture){
-        stream->cap_bufsize_frames = muldiv(params->duration, stream->dev_desc.mSampleRate, 10000000);
-        stream->cap_buffer = malloc(stream->cap_bufsize_frames * stream->fmt->nBlockAlign);
-    }
     params->result = S_OK;
 
 end:
     if(FAILED(params->result)){
         if(stream->converter) AudioConverterDispose(stream->converter);
         if(stream->unit) AudioComponentInstanceDispose(stream->unit);
+        free(stream->resamp_buffer);
+        free(stream->wrap_buffer);
+        free(stream->cap_buffer);
+        if(stream->local_buffer){
+            size = 0;
+            NtFreeVirtualMemory(GetCurrentProcess(), (void **)&stream->local_buffer,
+                                &size, MEM_RELEASE);
+        }
         free(stream->fmt);
         free(stream);
     } else {
@@ -1256,14 +1316,8 @@
     data->mBuffers[0].mDataByteSize = *nframes * stream->fmt->nBlockAlign;
     data->mBuffers[0].mNumberChannels = stream->fmt->nChannels;
 
-    if(stream->cap_offs_frames + *nframes > stream->cap_bufsize_frames){
+    if(*nframes > stream->cap_bufsize_frames - stream->cap_offs_frames){
         UINT32 chunk_frames = stream->cap_bufsize_frames - stream->cap_offs_frames;
-
-        if(stream->wrap_bufsize_frames < *nframes){
-            free(stream->wrap_buffer);
-            stream->wrap_buffer = malloc(data->mBuffers[0].mDataByteSize);
-            stream->wrap_bufsize_frames = *nframes;
-        }
 
         memcpy(stream->wrap_buffer, stream->cap_buffer + stream->cap_offs_frames * stream->fmt->nBlockAlign,
                chunk_frames * stream->fmt->nBlockAlign);
@@ -1274,8 +1328,11 @@
     }else
         data->mBuffers[0].mData = stream->cap_buffer + stream->cap_offs_frames * stream->fmt->nBlockAlign;
 
-    stream->cap_offs_frames += *nframes;
-    stream->cap_offs_frames %= stream->cap_bufsize_frames;
+    if(*nframes >= stream->cap_bufsize_frames - stream->cap_offs_frames)
+        stream->cap_offs_frames = *nframes -
+                                  (stream->cap_bufsize_frames - stream->cap_offs_frames);
+    else
+        stream->cap_offs_frames += *nframes;
     stream->cap_held_frames -= *nframes;
 
     if(packets)
@@ -1300,12 +1357,6 @@
         converted_list.mNumberBuffers = 1;
         converted_list.mBuffers[0].mNumberChannels = stream->fmt->nChannels;
         converted_list.mBuffers[0].mDataByteSize = wanted_frames * stream->fmt->nBlockAlign;
-
-        if(stream->resamp_bufsize_frames < wanted_frames){
-            free(stream->resamp_buffer);
-            stream->resamp_buffer = malloc(converted_list.mBuffers[0].mDataByteSize);
-            stream->resamp_bufsize_frames = wanted_frames;
-        }
 
         converted_list.mBuffers[0].mData = stream->resamp_buffer;
 
@@ -1497,7 +1548,8 @@
     }
 
     os_unfair_lock_unlock(&stream->lock);
-    if (!stream->timer_thread) create_unix_thread( &stream->timer_thread, name, unix_timer_loop, stream );
+    if(params->result == S_OK && (stream->flags & AUDCLNT_STREAMFLAGS_EVENTCALLBACK) && !stream->timer_thread)
+        create_unix_thread(&stream->timer_thread, name, unix_timer_loop, stream);
 
     return STATUS_SUCCESS;
 }
